Update Settings page to include "Other" section

delawski opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #5578

This PR removes the Mobile redirection toggle from the Onboarding Wizard and adds it to a new "Other" section on the AMP Settings page. The "Other" section also contains a Dev Tools toggle:

![Screenshot 2021-08-20 at 14 56 49](https://user-images.githubusercontent.com/478735/130236595-2644180d-6f62-4723-b302-198ba87a3aae.png)

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Remove mobile redirect toggle from Onboarding Wizard

delawski committed:
Add mobile redirection toggle to new Other Settings section

delawski committed:
Move `UserContextProvider` to general components directory

delawski committed:
Add `User` context and Dev Tools section to Settings page

delawski committed:
Fix e2e test

delawski committed:
Open "Other" panel before searching for mobile redirect toggle

delawski committed:
Wait for "Other" panel before attempting to click it

delawski commented:
I've been trying to fix the failing E2E test but ran out of time. I'm pretty sure there's an easy fix for that.

pierlon commented:
> I've been trying to fix the failing E2E test but ran out of time. I'm pretty sure there's an easy fix for that.

No worries. I'll take a look.

delawski commented:
> No worries. I'll take a look.

Thanks so much, @pierlon!

jwold committed:
Updated headings within the Other section of Settings

Per the comment from Weston on https://github.com/ampproject/amp-wp/issues/5578#issuecomment-890089314, I'm suggesting a change to the new "Other" panel. The goal is that the design should now look like this: https://d.pr/i/CPtRce. Feedback is welcome, haven't submitted a PR in a while!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=6501) <br/>All committers have signed the CLA.

jwold committed:
Update developer-tools.js

Saw an error to properly indent. Trying again.

delawski committed:
Remove unused import

delawski committed:
Update snapshots

delawski committed:
Add unit test for loading state

delawski committed:
Fix Other settings box spacing

delawski committed:
Fix E2E test

delawski committed:
Add E2E test for dev tools toggle

delawski commented:
@jwold Thanks for updating the component!

@pierlon @westonruter I've been able to fix the E2E tests, increase the test coverage, and also resolve the merge conflicts. Now the PR is "green" and I think it's good to go through the final review.

pierlon commented on assets/src/components/dev-tools-toggle/index.js:
```diff
@@ -0,0 +1,34 @@
+ * Internal dependencies
+ */
+import { AMPSettingToggle } from '../amp-setting-toggle';
+import { User } from '../../components/user-context-provider';
```
Import can be shortened.

```suggestion
import { User } from '../user-context-provider';
```

pierlon commented on assets/src/components/dev-tools-toggle/index.js:
```diff
@@ -0,0 +1,34 @@
+	return (
+		<AMPSettingToggle
+			checked={ true === developerToolsOption }
+			title={ __( 'Enable Dev Tools', 'amp' ) }
```
I don't think there's a need to use the shortened "Dev" form of "Developer", it may also cause confusion for some.

```suggestion
			title={ __( 'Enable Developer Tools', 'amp' ) }
```

pierlon commented on assets/src/settings-page/developer-tools.js:
```diff
@@ -0,0 +1,23 @@
+		<section className="developer-tools">
+			<DevToolsToggle />
+			<p>
+				{ __( 'Only enabled for your user account. This is not a sitewide setting.', 'amp' ) }
```
We should also mention what it's useful for, similar to how it was on the User Profile page:

<img width="910" alt="image" src="https://user-images.githubusercontent.com/16200219/130085256-3c2dc50f-d51b-413f-90dd-5693f2e227ff.png">

delawski replied:
How about something like this?

![Screenshot 2021-08-19 at 18 30 41](https://user-images.githubusercontent.com/478735/130107275-70a128b4-a42e-429c-9eb4-8a5ed881f6da.png)

delawski replied:
This is addressed in 753175f

pierlon commented on assets/src/components/unsaved-changes-warning/index.js:
```diff
@@ -12,7 +12,7 @@ import { __ } from '@wordpress/i18n';
  * Internal dependencies
  */
-import { User } from '../../onboarding-wizard/components/user-context-provider';
+import { User } from '../../components/user-context-provider';
```
Import can be shortened.

```suggestion
import { User } from '../user-context-provider';
```

delawski committed:
Apply suggestions from code review

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

delawski committed:
Update Jest snapshot

delawski committed:
Improve dev tools toggle description

delawski committed:
Allow user-related changes on Options page

pierlon commented on tests/e2e/specs/admin/other-settings.js:
```diff
@@ -0,0 +1,73 @@
+	it( 'persists the mobile redirect setting value', async () => {
+		const toggleSelector = '#other-settings .mobile-redirection .amp-setting-toggle input[type="checkbox"]';
+
+		// Confirm the mobile redirection is initially disabled.
```
```suggestion
		// Confirm the mobile redirection is initially enabled.
```

pierlon commented on assets/src/components/user-context-provider/index.js:
```diff
@@ -13,20 +13,22 @@ import PropTypes from 'prop-types';
- * @param {string} props.userRestPath             REST endpoint to retrieve options.
+ * @param {Object}  props                           Component props.
+ * @param {?any}    props.children                  Component children.
+ * @param {boolean} props.allowConfiguredPluginOnly Provide only if the plugin has been configured (Onboarding Wizard complete).
```
```suggestion
 * @param {boolean} props.allowConfiguredPluginOnly Provided only for functionality that requires the plugin to be configured.
```

pierlon committed:
Fix typo

pierlon committed:
Update param description

pierlon reviewed (approved):
Great work @delawski. Passing onto @westonruter for a final review.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5578-add-new-settings

* 'develop' of github.com:ampproject/amp-wp: (44 commits)
  Make placeholder dependency on video ID more explicit
  Fix adding start to placeholder link; improve coverage
  Let placeholder link point to YouTube permalink and not iframe src
  Prevent generating closing IMG tags
  Use coversDefaultClass
  Remove condition which cannot be reached since video ID is always provided
  Opt to pass-through iframe for unrecognized YouTube URL rather than convert to link
  Remove unreachable code
  Give more explicit array phpdoc type
  Fix issues identified by phpstan
  Try installing phpstan via shivammathur/setup-php
  Try downloading a specific version of phpstan instead of latest
  Fix yml syntax error
  Try removing phpstan before composer install
  Print phpstan version when running on GHA
  Revert "Revert phpstan ignores"
  Try adding phpstan version to composer cache key
  Revert phpstan ignores
  update unit test cases
  Sanitize attribute name Use original YouTube URL as placeholder when sanitizing raw embed Fix phpcs alignment warning Use constant instead of member variable Let Document::fromNode() obtain the ownerDocument Deduplicate iframe_props into constant
  ...

westonruter commented on assets/src/settings-page/developer-tools.js:
```diff
@@ -0,0 +1,26 @@
+				{ __( 'Enable AMP developer tools to surface validation errors when editing posts and viewing the site.', 'amp' ) }
+			</p>
+			<p>
+				{ __( 'Only enabled for your user account. This is not a sitewide setting. This presumes you have some experience coding with HTML, CSS, JS, and PHP.', 'amp' ) }
```
```suggestion
				{ __( 'This is a per-user setting. It presumes you have some experience coding with HTML, CSS, JS, and PHP. Once enabled you will have access to Validated URLs and Error Index in the admin menu, the Validate URL item in the admin bar, and the AMP Validation sidebar in the editor.', 'amp' ) }
```

westonruter commented on assets/src/components/user-context-provider/index.js:
```diff
@@ -13,20 +13,22 @@ import PropTypes from 'prop-types';
- * @param {string} props.userRestPath             REST endpoint to retrieve options.
+ * @param {Object}  props                           Component props.
+ * @param {?any}    props.children                  Component children.
+ * @param {boolean} props.allowConfiguredPluginOnly Provided only for functionality that requires the plugin to be configured.
```
This prop name and description are unclear to me. Can this be clarified?

delawski replied:
Yes, the description was totally wrong and the prop name could be better too.

In 418ec3c I've changed those. I'm still not convinced it's clear enough but maybe you can find a better wording?

To make it clearer, I've also replaced the REST _route_ we used: `/wp/v2/users/me` with a general `users` _path_: `/wp/v2/users`. This way, the prop name (`usersResourceRestPath`) should now semantically match its value.

What do you think?

westonruter replied:
I don't see that `allowConfiguredPluginOnly` was changed.

The users REST path change looks good.

delawski replied:
Oh my! I have no idea why I looked at the wrong prop (and "fixed" it). I'll get the `allowConfiguredPluginOnly` clarified.

delawski replied:
So this flag is used to allow fetching users data even if the plugin is not fully configured (the Onboarding Wizard has not been completed).

So far, the user data was fetched only after completing the Wizard but right now we need the user data earlier (it's possible to visit the AMP Settings page without going through the Wizard).

Naming is hard. Anything I come up with so far is long and may sound a bit awkward: `skipFetchingIfUnconfiguredPlugin`, `onlyFetchIfPluginIsConfigured`... Suggestions are well appreciated.

westonruter replied:
`onlyFetchIfPluginIsConfigured` sounds good to me!

delawski replied:
Perfect! It's been updated in 4b3a89a.

westonruter committed:
Add missing tests for OptionsMenu

westonruter committed:
Enable mobile redirection by default

westonruter committed:
Set current screen to fix test_enqueue_assets_right_hook_suffix

westonruter reviewed (commented):
I added long-missing tests for `OptionsMenu` which I've added in 121ca24. 

One of the requirements was:

> Remove "redirect mobile visitors to AMP" toggle from onboarding

I've implemented this in ba02cab.

The main thing left is the E2E tests being fixed, aside from the minor comments I'm leaving here.

westonruter commented:
Oh, hold on, E2E tests are passing now! 🤷 

Maybe they'll fail again when the above suggestions are applied.

delawski committed:
Update assets/src/settings-page/developer-tools.js

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Update `users` resource prop and its description

delawski committed:
Fix the unit test

delawski commented:
> Oh, hold on, E2E tests are passing now! 🤷

> Maybe they'll fail again when the above suggestions are applied.

That's strange. I don't seem to be getting the E2E error locally. I think this may be some kind of hazard where the Puppeteer doesn't wait for a specific selector before assessing its existence.

delawski committed:
Try fixing failing E2E test

delawski committed:
Improve prop name and description
